flock_unix.go

```diff
@@ -141,6 +150,37 @@ func (f *Flock) try(locked *bool, flag int) (bool, error) {
 		*locked = true
 		return true, nil
 	}
+	if !retried && f.retryError(err) {
+		retried = true
+		goto retry
+	}
 
 	return false, err
 }
+
+// retryError determines whether we should reopen the file handle in readwrite
+// mode and try again. This comes from util-linux/sys-utils/flock.c:
+//  Since Linux 3.4 (commit 55725513)
+//  Probably NFSv4 where flock() is emulated by fcntl().
+func (f *Flock) retryError(err error) bool {
```

Comment:
I'm not a huge fan of this method name as it doesn't match what it does. The method itself doesn't retry anything, and instead determines if we should retry and silently changes our file descriptor out from under us.

I'm not sure how I feel about the `os.OpenFile` error being silently swallowed by this method, since that would indicate something really weird is going on. In that case, I'd want to bubble the error up to the consumer because the validation you've done already indicates that it should _not_ fail.

How do you feel about something like `reopenFDOnError`. It could return a `bool` or an `error`. If there's an `error`, something went wrong and we should tell the caller. If bool is `false`, we decided not to reopen the file. If it's `true`, we should retry.